src/test/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageTest.java

```diff
@@ -0,0 +1,300 @@
+package org.jenkinsci.plugins.workflow.pipelinegraphanalysis;
+
+import junit.framework.Assert;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.graphanalysis.FlowChunkWithContext;
+import org.jenkinsci.plugins.workflow.graphanalysis.ForkScanner;
+import org.jenkinsci.plugins.workflow.graphanalysis.MemoryFlowChunk;
+import org.jenkinsci.plugins.workflow.graphanalysis.StandardChunkVisitor;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+
+/**
+ * Tests the stage collection
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
```

Comment:
In order to not clutter this PR up with extraneous changes, that would be a separate PR though.